File: typescript/Behavioral/20-Strategy/Context/BillingService.ts
import type { IPricingStrategy } from "../Strategy/IPricingStrategy";
import type { JobCard } from "../Model/JobCard";

export class BillingService {
  constructor(private strategy: IPricingStrategy) {}

  // Strategy can be swapped at runtime
  setStrategy(strategy: IPricingStrategy): void {
    console.log(`  [Billing] Pricing strategy set to: ${strategy.name}`);
    this.strategy = strategy;
  }

  generateInvoice(job: JobCard): number {
    const amount = this.strategy.calculate(job);

    console.log(`\n  Invoice for ${job}`);
    console.log(`    Strategy : ${this.strategy.name}`);
    console.log(`    Basis    : ${this.strategy.explain(job)}`);
    console.log(`    TOTAL    : N$${amount.toFixed(2)}`);

    return amount;
  }
}

File: csharp/Behavioral/20-Strategy/Context/BillingService.cs
namespace Strategy;

public class BillingService
{
    private IPricingStrategy _strategy;

    public BillingService(IPricingStrategy strategy)
    {
        _strategy = strategy;
    }

    // Strategy can be swapped at runtime
    public void SetStrategy(IPricingStrategy strategy)
    {
        Console.WriteLine($"  [Billing] Pricing strategy set to: {strategy.Name}");
        _strategy = strategy;
    }

    public decimal GenerateInvoice(JobCard job)
    {
        var amount = _strategy.Calculate(job);

        Console.WriteLine($"\n  Invoice for {job}");
        Console.WriteLine($"    Strategy : {_strategy.Name}");
        Console.WriteLine($"    Basis    : {_strategy.Explain(job)}");
        Console.WriteLine($"    TOTAL    : N${amount:N2}");

        return amount;
    }
}
